odoo/addons/base/transpiler/sourcemap_generator.py

```diff
@@ -0,0 +1,168 @@
+    browser to read.
+
+    This implementation of the SourceMapGenerator is a translation and adaptation of this implementation
+    in js https://github.com/mozilla/source-map. For performance purposes, we have removed all unnecessary
+    functions/steps for our use case. This simpler version does a line by line mapping, with the ability to
+    add offsets at the start and end of a file. (when we have to add comments on top a transpiled file by example).
+    """
+    def __init__(self, **kwargs):
+        self._file = kwargs.get("file")
+        self._sourceRoot = kwargs.get("sourceRoot")
+        self._sources = {}
+        self._mappings = []
+        self._sourcesContents = {}
+        self._version = 3
+        self.cache = {}
+
+    def _serialize_mappings(self):
+        """
+        A source map mapping is encoded with the base 64 VLQ format.
+        This function encodes the readable source to the format.
+
+        :return the encoded content
+        """
+        previousGeneratedLine = 1
+        previousOriginalLine = 0
+        previousSource = 0
+        encoded_column = base64vlq_encode(0)
+        result = ""
+        for mapping in self._mappings:
+            if (mapping["generatedLine"] != previousGeneratedLine):
+                while (mapping["generatedLine"] != previousGeneratedLine):
+                    result += ";"
+                    previousGeneratedLine += 1
+
+            if (mapping["source"] is not None):
+                sourceIdx = self._sources[mapping["source"]]
+                source = sourceIdx - previousSource
+                previousSource = sourceIdx
+
+                # lines are stored 0-based in SourceMap spec version 3
+                line = mapping["originalLine"] - 1 - previousOriginalLine
+                previousOriginalLine = mapping["originalLine"] - 1
+
+            if (source, line) not in self.cache:
+                self.cache[(source, line)] = "".join([encoded_column, base64vlq_encode(source), base64vlq_encode(line), encoded_column])
+            result += self.cache[source, line]
+        return result
+
+    def to_json(self):
+        """
+        Generates the json sourcemap.
+        It is the main function that assembles all the pieces.
+
+        :return {str} valid sourcemap in json format
+        """
+        mapping = {
+            "version": self._version,
+            "sources": list(self._sources.keys()),
+            "mappings": self._serialize_mappings(),
+            "sourcesContent": [self._sourcesContents[source] for source in self._sources]
+        }
+        if (self._file is not None):
+            mapping["file"] = self._file
+
+        if (self._sourceRoot is not None):
+            mapping["sourceRoot"] = self._sourceRoot
+
+        return mapping
+
+    def add_source(self, source_name, source_content, previous_generated_line, nb_skip_line_output=0):
```

Comment:
I find this API very awkward, because the state of the sourcemap generator is partly outside the object, i.e., `previous_generated_line`.  Isn't there a way to make this cleaner?

Reply:
@FrancoisGe Can you look at this one ? See if your API can be better ?